src/term/mod.rs

```diff
@@ -299,6 +300,101 @@ pub struct Cursor {
     charsets: Charsets,
 }
 
+pub struct VisualBell {
+    /// Visual bell animation
+    animation: VisualBellAnimation,
+
+    /// Visual bell duration
+    duration: Duration,
+
+    /// The last time the visual bell rang, if at all
+    start_time: Option<Instant>,
+}
+
+fn cubic_bezier(p0: f64, p1: f64, p2: f64, p3: f64, x: f64) -> f64 {
+    (1.0 - x).powi(3) * p0 +
+    3.0 * (1.0 - x).powi(2) * x * p1 +
+    3.0 * (1.0 - x) * x.powi(2) * p2 +
+    x.powi(3) * p3
+}
+
+impl VisualBell {
+    pub fn new(config: &VisualBellConfig) -> VisualBell {
+        VisualBell {
+            animation: config.animation(),
+            duration: config.duration(),
+            start_time: None,
+        }
+    }
+
+    /// Ring the visual bell, and return its intensity.
+    pub fn ring(&mut self) -> f64 {
+        let now = Instant::now();
+        self.start_time = Some(now);
+        self.intensity_at_instant(now)
+    }
+
+    /// Get the currenty intensity of the visual bell. The bell's intensity
+    /// ramps down from 1.0 to 0.0 at a rate determined by the bell's duration.
+    pub fn intensity(&self) -> f64 {
+        self.intensity_at_instant(Instant::now())
+    }
+
+    /// Check whether or not the visual bell has completed "ringing".
+    pub fn completed(&self) -> bool {
+        match self.start_time {
+            Some(earlier) => Instant::now().duration_since(earlier) > self.duration,
+            None => true
+        }
+    }
+
+    /// Get the intensity of the visual bell at a particular instant. The bell's
+    /// intensity ramps down from 1.0 to 0.0 at a rate determined by the bell's
+    /// duration.
+    pub fn intensity_at_instant(&self, instant: Instant) -> f64 {
+        let x = if self.duration == Duration::from_secs(0) {
+            1.0
```

Comment:
Yeah, I can see how this reads strangely. I've refactored it a bit, renamed some variables, and added comments. Hopefully it will be clearer.